src/context.rs

```diff
@@ -30,19 +31,83 @@ impl<'a> QueryMatchContext<'a> {
         file_contents: &'a [u8],
         rule: &'a ResolvedRule,
         reported_any_violations: &'a AtomicBool,
+        context: &'a Context,
     ) -> Self {
         Self {
             path,
             file_contents,
             rule,
             reported_any_violations,
+            context,
         }
     }
 
     pub fn report(&self, violation: Violation) {
         self.reported_any_violations.store(true, Ordering::Relaxed);
         print_violation(&violation, self);
     }
+
+    pub fn get_node_text(&self, node: Node) -> &str {
+        node.utf8_text(self.file_contents).unwrap()
+    }
+
+    pub fn maybe_get_single_matching_node_for_query<'query, 'enclosing_node>(
+        &self,
+        query: impl Into<ParsedOrUnparsedQuery<'query>>,
+        enclosing_node: Node<'enclosing_node>,
+    ) -> Option<Node<'enclosing_node>> {
+        let query = query.into().into_parsed(self.context.language);
+        let mut query_cursor = QueryCursor::new();
+        let mut matches = query_cursor.matches(&query, enclosing_node, self.file_contents);
+        let first_match = matches.next()?;
+        if matches.next().is_some() {
+            return None;
+        }
+        let mut nodes_for_default_capture_index = first_match.nodes_for_capture_index(0);
+        let first_node = nodes_for_default_capture_index.next()?;
+        if nodes_for_default_capture_index.next().is_some() {
+            return None;
+        }
+        Some(first_node)
+    }
+
+    pub fn get_number_of_query_matches<'query, 'enclosing_node>(
+        &self,
+        query: impl Into<ParsedOrUnparsedQuery<'query>>,
+        enclosing_node: Node<'enclosing_node>,
+    ) -> usize {
+        let query = query.into().into_parsed(self.context.language);
+        let mut query_cursor = QueryCursor::new();
+        query_cursor
+            .matches(&query, enclosing_node, self.file_contents)
+            .count()
+    }
+}
+
+pub enum ParsedOrUnparsedQuery<'query_text> {
```

Comment:
I think this is "unused" at the moment (ie I'm always passing a query string rather than a pre-parsed query to these helper methods) but it felt like the right type for it to accept